ls1intum/Artemis
# `Plagiarism checks`: Mark submissions detected by cpc as plagiarism and reduce their score

# ⚠️ This PR contains braking db changes. Please use ts4 to deploy it. More notes in _Steps for Testing_

### Checklist


#### General
#### Server
#### Client
#### Changes affecting Programming Exercises
_nothing_

### Motivation and Context
My previous PR #6666 introduced cpc, but until this PR there was no immediate feedback delivered to the student about plagiarism detected in his or her submission. Changes introduces here notify the student about suspected plagiarism.

Due to issue with the current implementation of handling results of plagiarism checks for programming exercises this PR enables new features only for text and modeling exercises. Change for programming exercises will be introduced in my next PR.

This implements CPC-3 and CPC-5 milestone. More about that can be discovered in my thesis proposal: https://confluence.ase.in.tum.de/display/ArTEMiS/MA+Jakub+Riegel?preview=/157431627/166332701/riegel_proposal.pdf

### Description
This PR uses changes from #6666 and should be merged after it.

Three features are introduced:
1. Student sees warning about plagiarism detected in his or her submission.
2. Score of that submission is reduced to 0.
3. After exercise deadline student can dispute score reduced by cpc.

**Warning about plagiarism detected**
On the submission page and editor (if applicable) the student sees a gentle note '⚠️ _Suspicion of plagiarism!_ '. This gives student a clear information that this is needs to be fixed, while not imposing any negative consequences.

**Score reduced to 0**
Submission with detected plagiarism receives a _Result_ with 0 score and a comment that this is due to detected plagiarism. Artemis treats and displays this result as any other results. 

~~For programming exercises submitting next submission will generate a new tests result which overrides plagiarism results. In case that new submission is also plagiarism the next cpc run will add 0 score again.~~

For text and modeling exercises submitting again doesn't remove 0 score automatically. It is removed during the next cpc run.

**Disputing cpc results**
Disputing cpc results is possible after the deadline of the exercise. Students use the same interface as for disputing assessment results. Instructors handle those disputes the same way they handle assessments disputes. 

### Steps for Testing
- 1 Instructor
- 2 Students
- 1 Course
- for convenience use the `cpc-mark-submissions-as-plagiarism-1min` [branch](https://github.com/ls1intum/Artemis/tree/cpc-mark-submissions-as-plagiarism-1min) which triggers checks every 5 minutes
- **due to database changes please deploy it to ts4 by using #6841** 

General:
1. Create programming, text or modelling exercise with continuous plagiarism control enabled (use _Enabled_ checkbox on exercise creation form)
4. Create two similar submission for the exercise
5. Wait 2 minutes (results need to synchronise)
6. Verify that ⚠️ _Suspicion of plagiarism_ warning was placed and 0 points result was posted (see screenshots at the bottom)
7. Change one submission to something completely different
8. Wait 2 minutes
9. Verify that warnings and results are deleted

Disputing results:
1. Generate submissions with plagiarism detected by the cpc (using the steps above)
2. Change exercise due date to the past
3. As a student go into your submission page and dispute the result
4. As instructor approve and decline the dispute
5. If approved: new scores are displayed; if not: old results are displayed 

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ContinuousPlagiarismControlService.java | 93% | ✅                           |
| ContinuousPlagiarismControlResultsService.java | 100% | ✅                           |

### Screenshots
#### Plagiarism detected warning
Exercises overview:
<img width="1006" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/72a78ebc-5c66-468a-9bde-45c589b71b10">

Exercise view:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/81887918-648d-41e0-a60a-f2708384a7fc">

#### Plagiarism score result
Overview:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/625328d1-bcac-495d-9846-2d51a47ab55d">

Details:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/32bbc47a-1a68-40fd-8596-f5f71ccad906">

Editor view (text & modeling):
<img width="1511" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/64d6ab07-9b80-4f4a-984c-cac5fc5fbdaf">
